### PR TITLE
Splitted CEP to separate repository

### PR DESCRIPTION
**Pull request by YoungFaithful**
- multiple years
- segmentation

### CODE REVIEW

File: docs/make.jl [hunk 1]
```diff
@@ -9,7 +9,7 @@ makedocs(sitename="ClustForOpt.jl",
         "Workflow" => "workflow.md",
         "Load Data" => "load_data.md",
         "Clustering" => "clust.md",
-        "Optimization" => ["opt.md", "opt_cep.md", "opt_cep_data.md"]
+        "Optimization" => "opt.md"
```
Comment (YoungFaithful) [inline docs/make.jl]: in other documentation

File: docs/src/index.md [hunk 2]
```diff
@@ -4,9 +4,9 @@
 [![](https://img.shields.io/badge/docs-dev-blue.svg)](https://holgerteichgraeber.github.io/ClustForOpt.jl/dev)
 [![Build Status](https://travis-ci.com/holgerteichgraeber/ClustForOpt.jl.svg?token=HRFemjSxM1NBCsbHGNDG&branch=master)](https://travis-ci.com/holgerteichgraeber/ClustForOpt.jl)
 
-[ClustForOpt](https://github.com/holgerteichgraeber/ClustForOpt.jl) is a [julia](https://www.juliaopt.com) implementation of clustering methods for finding representative periods for the optimization of energy systems. The package furthermore provides a multi-node capacity expansion model.
+[ClustForOpt](https://github.com/holgerteichgraeber/ClustForOpt.jl) is a [julia](https://www.juliaopt.com) implementation of clustering methods for finding representative periods for optimization problems. A utilization in a scalable capacity expansion problem can be found in the package [CEP](https://github.com/YoungFaithful/CEP.jl).
```
Comment (YoungFaithful) [inline docs/src/index.md]: linking the other

File: docs/src/load_data.md [hunk 3]
```diff
@@ -20,28 +18,9 @@ load_timeseries_data
 using ClustForOpt
 state="GER_1"
 # laod ts-input-data
-ts_input_data, = load_timeseries_data("CEP", state; K=365, T=24)
+ts_input_data = load_timeseries_data("CEP", state; K=365, T=24)
```
Comment (YoungFaithful) [inline docs/src/load_data.md]: - load_timeseries_data has only one output left
- we have to discuss how we shall do the data handling, examples

File: examples/workflow_introduction.jl [hunk 4]
```diff
@@ -9,7 +9,7 @@ The data is for a Capacity Expansion Problem "CEP"
 and for the single node representation of Germany "GER_1"
 The original timeseries has 8760 entries (one for each hour of the year)
 It should be cut into K=365 periods (365 days) with T=24 timesteps per period (24h per day) =#
-ts_input_data, = load_timeseries_data("CEP", "GER_1"; K=365, T=24)
+ts_input_data = load_timeseries_data(normpath(joinpath(dirname(@__FILE__),"TS_GER_1")); T=24, years=[2016])
```
Comment (YoungFaithful) [inline examples/workflow_introduction.jl]: Different location of data in `examples\TS_GER_1`

File: src/ClustForOpt.jl [hunk 5]
```diff
@@ -7,27 +7,25 @@
  #####################
 module ClustForOpt
   using Reexport
-  @reexport using StatsKit
-  @reexport using JLD2
+  using StatsKit
   @reexport using FileIO
-  using JuMP
+  using JuMP #QUESTION Part of ClustForOpt?
```
Comment (YoungFaithful) [inline src/ClustForOpt.jl]: is that used for k-medoids or so?

File: src/ClustForOpt.jl [hunk 6]
```diff
@@ -7,27 +7,25 @@
...
 
    #TODO how to make PyPlot, PyCall, and TimeWarp optional? -> only import when needed
 
    export run_opt,
           run_clust,
+          run_darkperiod_clust,
+          run_pure_clust,
```
Comment (YoungFaithful) [inline src/ClustForOpt.jl]: Shall I remove that? more research code, I suppose.

Comment (YoungFaithful) [reply src/ClustForOpt.jl]: pure is for only one attribute clustering
darkperiod is for week or so

File: src/ClustForOpt.jl [hunk 7]
```diff
@@ -7,27 +7,25 @@
...
           get_sup_kw_args,
           InputData,
           FullInputData,
           ClustData,
           ClustDataMerged,
+          ClustResult,
```
Comment (YoungFaithful) [inline src/ClustForOpt.jl]: need to export abstract type for reference in other package

File: src/ClustForOpt.jl [hunk 8]
```diff
@@ -44,22 +42,16 @@ module ClustForOpt
...
           resize_medoids,
           load_timeseries_data,
-          load_cep_data,
-          get_cep_variable_value,
-          get_cep_variable_set,
-          get_cep_slack_variables,
-          get_cep_design_variables,
-          get_total_demand
-
+          combine_timeseries_weather_data
```
Comment (YoungFaithful) [inline src/ClustForOpt.jl]: Allow one year of demand to be joint with multiple years of weather data

File: src/ClustForOpt.jl [hunk 9]
```diff
@@ -44,22 +42,16 @@ module ClustForOpt
...
   include(joinpath("utils","load_data.jl"))
   include(joinpath("optim_problems","run_opt.jl"))
-  include(joinpath("optim_problems","opt_cep.jl"))
   include(joinpath("clustering","run_clust.jl"))
   include(joinpath("clustering","exact_kmedoids.jl"))
   include(joinpath("clustering","extreme_vals.jl"))
   include(joinpath("clustering","attribute_weighting.jl"))
-
+  include(joinpath("clustering","intraperiod_segmentation.jl"))
+  include(joinpath("clustering","other_clust.jl"))
```
Comment (YoungFaithful) [inline src/ClustForOpt.jl]: "Research part" - Shall we remove it?

File: src/clustering/exact_kmedoids.jl [hunk 10]
```diff
@@ -31,11 +31,11 @@ N_i = size(data,2)
 
 
 # calculate distance matrix
-d_mat=pairwise(_dist,data)
+d_mat=pairwise(_dist,data; dims=2)
```
Comment (YoungFaithful) [inline src/clustering/exact_kmedoids.jl]: required expl.

File: src/clustering/extreme_vals.jl [hunk 11]
```diff
@@ -78,39 +79,55 @@ identifies a single simple extreme value from the data and returns column index
...
-  !(data_type in keys(data.data)) && @error("the provided data type - "*data_type*" - is not contained in data")
-  return simple_extr_val_ident(data.data[data_type]; extremum=extremum, peak_def=peak_def)
+                               peak_def::String="absolute",
+                               periods::Int64=1)
+  for name in keys(data.data)
+    attr=split(name,"-")[1]
+    if name==data_type
+      return simple_extr_val_ident(data.data[data_type]; extremum=extremum, peak_def=peak_def, periods=periods)
+    elseif attr==data_type
+      return simple_extr_val_ident(data.data[name]; extremum=extremum, peak_def=peak_def, periods=periods)
```
Comment (YoungFaithful) [inline src/clustering/extreme_vals.jl]: can either be an attribute "wind" or specific name "wind-dena81"

File: src/clustering/extreme_vals.jl [hunk 12]
```diff
@@ -65,10 +65,11 @@ identifies a single simple extreme value from the data and returns column index
 data_type: any attribute from the attributes contained within *data*
 extremum: "min" or "max"
 peak_def: "absolute" or "integral"
+periods: number of periods combined to analyze
 """
 function simple_extr_val_ident(data::ClustData,
                                extreme_value_descr::SimpleExtremeValueDescr)
-  return simple_extr_val_ident(data, extreme_value_descr.data_type; extremum=extreme_value_descr.extremum, peak_def=extreme_value_descr.peak_def)
+  return simple_extr_val_ident(data, extreme_value_descr.data_type; extremum=extreme_value_descr.extremum, peak_def=extreme_value_descr.peak_def, periods=extreme_value_descr.periods)
```
Comment (YoungFaithful) [inline src/clustering/extreme_vals.jl]: multiple periods can be combined (multiple days e.g.)

File: src/clustering/extreme_vals.jl [hunk 13]
```diff
@@ -78,39 +79,55 @@ identifies a single simple extreme value from the data and returns column index
...
   if extremum=="max"
-    idx = findmax(data_eval)[2][2]
+    idx_k = findmax(data_eval)[2][2]
   elseif extremum=="min"
-    idx = findmin(data_eval)[2][2]
+    idx_k = findmin(data_eval)[2][2]
   else
     @error("extremum - "*extremum*" - not defined")
   end
+  idx=collect(idx_k:(idx_k+delta_period))
```
Comment (YoungFaithful) [inline src/clustering/extreme_vals.jl]: allowing multiple extreme value selections on same data

File: src/clustering/extreme_vals.jl [hunk 14]
```diff
@@ -168,7 +194,16 @@ function extreme_val_output(data::ClustData,
...
-  extr_vals = ClustData(data.region,K_ed,data.T,data_ed,weights_ed;mean=data.mean,sdv=data.sdv)
+  deltas_ed=data.deltas[:,unique_extr_val_idcs]
+  #QUESTION What with k_ids?
+  # if original time series period isn't represented by any extreme period it has value 0
+  k_ids_ed=zeros(Int64,size(data.k_ids))
+  index_k_ids_data=findall(data.k_ids.!=0)
+  k_ids_ed_data=k_ids_ed[index_k_ids_data]
+  # each original time series period which is represented recieves the number of it's extreme period in this extreme value output
+  k_ids_ed_data[unique_extr_val_idcs]=collect(1:K_ed)
+  k_ids_ed[index_k_ids_data]=k_ids_ed_data
```
Comment (YoungFaithful) [inline src/clustering/extreme_vals.jl]: k_ids are included into the ClustData and need to be adjusted (to 0) if Extreme is selected.

File: src/clustering/run_clust.jl [hunk 15]
```diff
@@ -27,9 +27,56 @@ function run_clust(
...
+    if !isempty(attribute_weights)
+      data_norm = attribute_weighting(data_norm,attribute_weights)
+    end
+    data_norm_merged = ClustDataMerged(data_norm)
+
+    #clustering
+    b_merged, cost, cost_best, iter =run_clust(data_norm_merged, data; method=method, representation=representation, n_clust=n_clust, n_init=n_init, iterations=iterations, orig_k_ids=deepcopy(data.k_ids), kwargs...)
 
+     if n_seg!=b_merged.T &&  n_seg!=0
+       b_merged=intraperiod_segmentation(b_merged;n_seg=n_seg,norm_scope=norm_scope,iterations=iterations)
```
Comment (YoungFaithful) [inline src/clustering/run_clust.jl]: allow segmentation

File: src/clustering/run_clust.jl [hunk 16]
```diff
@@ -27,9 +27,56 @@ function run_clust(
     # normalize
     # TODO: implement 0-1 normalization and add as a choice to runclust
     data_norm = z_normalize(data;scope=norm_scope)
-    data_norm_att = attribute_weighting(data_norm,attribute_weights)
-    data_norm_merged = ClustDataMerged(data_norm_att)
+    if !isempty(attribute_weights)
+      data_norm = attribute_weighting(data_norm,attribute_weights)
```
Comment (YoungFaithful) [inline src/clustering/run_clust.jl]: only if not empty

File: src/clustering/run_clust.jl [hunk 17]
```diff
@@ -27,9 +27,56 @@ function run_clust(
     # normalize
     # TODO: implement 0-1 normalization and add as a choice to runclust
     data_norm = z_normalize(data;scope=norm_scope)
-    data_norm_att = attribute_weighting(data_norm,attribute_weights)
-    data_norm_merged = ClustDataMerged(data_norm_att)
```
Comment (YoungFaithful) [inline src/clustering/run_clust.jl]: split into different function

File: src/clustering/run_clust.jl [hunk 18]
```diff
@@ -178,7 +250,7 @@ function run_clust_kmeans_centroid(
         centers_norm = mean(data_norm.data,dims=2) # should be 0 due to normalization
         clustids = ones(Int,size(data_norm.data,2))
         centers = undo_z_normalize(centers_norm,data_norm.mean,data_norm.sdv;idx=clustids) # need to provide idx in case that sequence-based normalization is used
-        cost = sum(pairwise(SqEuclidean(),centers_norm,data_norm.data)) #same as sum((seq_norm-repmat(mean(seq_norm,2),1,size(seq,2))).^2)
+        cost = sum(pairwise(SqEuclidean(),centers_norm,data_norm.data; dims=2)) #same as sum((seq_norm-repmat(mean(seq_norm,2),1,size(seq,2))).^2)
```
Comment (YoungFaithful) [inline src/clustering/run_clust.jl]: explicit

File: src/clustering/run_clust.jl [hunk 19]
```diff
@@ -343,7 +415,7 @@ function run_clust_hierarchical_centroid(
     iterations::Int;
     _dist::SemiMetric = SqEuclidean()
     )
-    ~,weights,clustids,~,iter= run_clust_hierarchical(data_norm,n_clust,iterations;_dist=_dist)
+    x,weights,clustids,x,iter= run_clust_hierarchical(data_norm.data,n_clust,iterations;_dist=_dist)
```
Comment (YoungFaithful) [inline src/clustering/run_clust.jl]: other symbol needed

File: src/utils/datastructs.jl [hunk 20]
```diff
@@ -7,6 +7,7 @@ abstract type ClustResult end
 "FullInputData"
 struct FullInputData <: TSData
  region::String
+ years::Array{Int64}
```
Comment (YoungFaithful) [inline src/utils/datastructs.jl]: mulit year

File: src/utils/datastructs.jl [hunk 21]
```diff
@@ -23,24 +24,30 @@ end
...
 struct ClustData <: TSData
  region::String
+ years::Array{Int64}
  K::Int
  T::Int
  data::Dict{String,Array}
  weights::Array{Float64}
  mean::Dict{String,Array}
  sdv::Dict{String,Array}
+ deltas::Array{Float64,2}
```
Comment (YoungFaithful) [inline src/utils/datastructs.jl]: segmentation or other length of input data possible

File: src/utils/datastructs.jl [hunk 22]
```diff
@@ -23,24 +24,30 @@ end
...
+ k_ids::Array{Int64}
```
Comment (YoungFaithful) [inline src/utils/datastructs.jl]: order of original time-series

**Comment by YoungFaithful** [comment]
The test of the master is still depending on `/tests/cep.jl` that doesn't exist any more.
We should figure out a new testing strategy

**Comment by holgerteichgraeber** [comment]
Changed PR base to `dev` instead of `master`

**Comment by holgerteichgraeber** [comment]
All the checks fail. What should we do to fix it?

File: docs/src/index.md (hunk 2 above)
Comment (holgerteichgraeber) [inline docs/src/index.md]:
```suggestion
[ClustForOpt](https://github.com/holgerteichgraeber/ClustForOpt.jl) is a [julia](https://www.juliaopt.com) implementation of clustering methods for finding representative periods for optimization problems. An example use case of a scalable capacity expansion problem can be found in the package [CEP](https://github.com/YoungFaithful/CEP.jl).
```

**Comment by YoungFaithful** [comment]
> All the checks fail. What should we do to fix it?

This is due to the fact, that testing the merge is still depending the travis.yml in the master branch, which tries to call /tests/cep.jl - this doesn't exist any more. We need simply ignore it and design new testing for ClustForOpt specifically.

File: src/utils/datastructs.jl [hunk 23]
```diff
@@ -321,27 +266,32 @@ function ClustData(data::FullInputData,
...
 function ClustDataMerged(region::String,
+                       years::Array{Int64,1},
                        K::Int,
                        T::Int,
                        data::Array,
                        data_type::Array{String},
-                       weights::Array{Float64};
+                       weights::Array{Float64},
+                       k_ids::Array{Int64,1};
+                       deltas::Array{Float64}=ones(T,K),
```
Comment (holgerteichgraeber) [inline src/utils/datastructs.jl]: why did you take out the optional arguments?

**Comment by holgerteichgraeber** [comment]
I'll go through the merge conflicts and get it ready for merge.

**Pull request merged**